Implement django-configurations

cadecairos opened the pull request:
~**Marking as a launch blocker because we need this to ensure we don't leak PII in server logs on production.**~

In order to better support differing configurations in dev/stage/prod I'm going to implement [django-configurations](https://django-configurations.readthedocs.io/en/stable/).

This is still WIP since I haven't checked the tests (100% chance they're failing) As of EOD Oct 30th I've just completed coding and a single successful startup of the app.

cadecairos commented:
@patjouk not quite ready yet, but let me know if you think this approach works for you.

cadecairos commented on donate/settings/base.py:
```diff
@@ -0,0 +1,386 @@
+
+    ALLOWED_HOSTS = env('ALLOWED_HOSTS')
+
+    @property
```
these `@property` decorated methods act as computed properties for django settings that need to be evaluated when accessed, or after class declaration.

cadecairos commented on donate/settings/base.py:
```diff
@@ -0,0 +1,386 @@
+    def CSP_WORKER_SRC(self):
+        return env('CSP_WORKER_SRC', default=self.CSP_DEFAULT)
+
+    @classmethod
```
django-configurations provides three hooks for configuration classes, `pre_setup`, `setup`, and `post_setup`, where you can add code that needs to execute during the loading of your django settings.

alanmoo commented:
@cadecairos There's some conflicts

alanmoo commented on README.md:
````diff
@@ -78,8 +78,22 @@ When the Django server is running, you can start the Django shell with:
+```python
+from donate.settings import Base
+
+class MyConfig(Base):
````
```suggestion

class MyConfig(Base):
```

alanmoo replied:
To make copy/paste into a new file not fail linting.

alanmoo commented on README.md:
```diff
@@ -78,8 +78,22 @@ When the Django server is running, you can start the Django shell with:
+
+class MyConfig(Base):
+    # Add django settings here
+    pass
```
```suggestion
    pass
    
```

cadecairos commented on donate/settings/staging.py:
```diff
@@ -0,0 +1,20 @@
+    @classmethod
+    def pre_setup(cls):
+        cls.LOGGING['loggers']['rq.worker']['handlers'] = ['DEBUG']
+        cls.LOGGING['loggers']['rq.worker']['level'] = 'DEBUG'
```
@alanmoo FYI this is what triggers worker queue logs on staging, but not on production

cadecairos commented:
> KeyError: "Couldn't setup configuration 'donate.settings.LocalConfig':   "

@alanmoo your `.env` should set `DJANGO_CONFIGURATION` to "LocalConfig" not "donate.settings.LocalConfig", and `DJANGO_SETTINGS_MODULE` to "donate.settings"

alanmoo commented on env.default:
```diff
@@ -1,10 +1,12 @@
 CORS_WHITELIST=*
 CORS_REGEX_WHITELIST=
-DATABASE_URL=
+DATABASE_URL=postgres://postgres@postgres:5432/donate
```
```suggestion
DATABASE_URL=postgres://donate:mozilla@postgres:5432/donate
```

cadecairos commented:
> I feel like the additional steps to customize a local deployment should be optional, and this should definitely still build without having to create a file that ends up getting git-ignored.

I agree. I think the default `Development` Configuration needs to be expanded to include the most commonly changed variable and ensure that they can be modified in the .env

I originally added the option for a Local.py to let devs have the option for "total control" over the configuration, but it really shouldn't be necessary to use it for day-to-day work.

alanmoo commented:
Sounds right. So should we do all of the following?

* Remove any references to local.py/LocalConfig
* Update everything in Development.py to check for the env var, or use the default that's currently on that line (for example, `ALLOWED_HOSTS = '*'` becomes `ALLOWED_HOSTS = env('ALLOWED_HOSTS', default='*')`)
* Remove the defaults from those values in env.default (just make it `ALLOWED_HOSTS=` in .env)

patjouk commented on donate/settings/development.py:
```diff
@@ -0,0 +1,28 @@
+    CSRF_COOKIE_SECURE = env('CSRF_COOKIE_SECURE')
+    SESSION_COOKIE_SECURE = env('SESSION_COOKIE_SECURE')
+
+    AWS_LOCATION = 'donate'
```
You might want to remove this if you don't use S3 for local dev

cadecairos replied:
yeah I'll leave this to the S3 configuration mixin

patjouk commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,68 @@
+# we error out first.
+env = environ.Env(
+    ALLOWED_HOSTS=(list, '*'),
+    ASSET_DOMAIN=(str, 'donate.mofoprod.net'),
```
It's not used on Heroku. Are we're using `AWS_S3_CUSTOM_DOMAIN` instead?

cadecairos replied:
Hmm I don't know what this is for. I'll dig a little and find out if we depend on it

cadecairos replied:
This looks liek a relic from the previous network-api repository merge.

patjouk commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,68 @@
+    AWS_REGION=(str, ''),
+    CONTENT_TYPE_NO_SNIFF=(bool, True),
+    CORS_REGEX_WHITELIST=(tuple, None),
+    CORS_WHITELIST=(tuple, ()),
```
I didn't saw those on Heroku either

patjouk commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,68 @@
+    USE_X_FORWARDED_FOR=(bool, False),
+    X_FRAME_OPTIONS=(str, 'DENY'),
+    XSS_PROTECTION=(bool, True),
+    REDIS_URL=(str, 'postgres://donate:mozilla@postgres:5432/donate'),
```
typo ;)

patjouk commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,68 @@
+    CONTENT_TYPE_NO_SNIFF=(bool, True),
+    CORS_REGEX_WHITELIST=(tuple, None),
+    CORS_WHITELIST=(tuple, ()),
+    DATABASE_URL=(str, 'postgres://donate:mozilla@postgres:5432/donate'),
```
Do we want the default to be `""` instead?

cadecairos replied:
Any reason in particular we'd want no default?

alanmoo commented on donate/settings/development.py:
```diff
@@ -0,0 +1,28 @@
+from .base import Base
+from .redis import Redis
+from .braintree import Braintree
+
```
I've never used S3 locally; I feel like leaving that in .env is the best approach as it feels like a fair amount of overhead to get it working locally when it's not needed.

cadecairos replied:
yeah, S3 locally is not a common dev use case, but one we should support, I'll double check to  make sure it's still possible.

cadecairos replied:
should be better now

cadecairos commented on donate/settings/base.py:
```diff
@@ -0,0 +1,389 @@
+    RANDOM_SEED = env('RANDOM_SEED')
+
+    # Initialize these attributes to None
+    BASKET_API_ROOT_URL = None
```
yeah there's no reason to define it this way. I'll update

cadecairos commented:
I've updated a bunch of things based on the feedback I've received so far. There have been other changes made when I found certain issues, such as how we've been using environ incorrectly this whole time, and the Thunderbird configuration classes in lieu of using a boolean flag.

Pomax commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,95 @@
+import environ
+
+CSP_DEFAULT_SRC = ("'self'")
+CSP_SCRIPT_SRC = ("'self'", '\'unsafe-inline\'', '\'unsafe-eval\'',
```
if `"'...'"` works, we should probably use that for the rest of the settings too, rather than `'\'...\''`.

Pomax commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,95 @@
+)
+
+
+app = environ.Path(__file__) - 2
```
is this... arithmetics on paths? O_O

cadecairos replied:
hahaha that was my reaction. but apparently it works. it's _magic_

alanmoo replied:
![skeptical face](https://media.giphy.com/media/600QaqOkDQnWo5Mau0/giphy.gif)

cadecairos replied:
I was curious, this is a `Path` class defined in django-environ: https://github.com/joke2k/django-environ/blob/develop/environ/environ.py#L735-L749

Pomax commented on donate/settings/production.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class ThunderbirdProduction(Production, ThunderbirdOverrides):
+    INSTALLED_APPS = ThunderbirdOverrides.INSTALLED_APPS + Production.INSTALLED_APPS
```
this is cool

Pomax commented on donate/settings/production.py:
```diff
@@ -0,0 +1,32 @@
+    @property
+    def TEMPLATES(self):
+        config = Production.TEMPLATES
+        config[0]['DIRS'] = ThunderbirdOverrides.TEMPLATES_DIR + config[0]['DIRS']
```
this is also cool

Pomax commented on donate/settings/staging.py:
```diff
@@ -0,0 +1,31 @@
+        cls.LOGGING['loggers']['rq.worker']['level'] = 'INFO'
+
+
+class ThunderbirdStaging(Staging, ThunderbirdOverrides):
```
Is there a way to make the ThunderbirdOverrides automatically hook into the `INSTALLED_APPS` and `TEMPLATES` variables? Or does it not have access to "the object that's been created so far" in a multiple inheritance chain?

cadecairos replied:
I initially tried that, but django-configurations doesn't seem to work that way. The solution I ended up with is the first combination of configurations that actually did what I was trying to do.

cadecairos commented:
@patjouk @alanmoo I believe I've addressed your reviews, could you both take another look so we can merge this and validate it in our various environments?

cadecairos commented on donate/settings/base.py:
```diff
@@ -0,0 +1,373 @@
+    BASKET_API_ROOT_URL = env('BASKET_API_ROOT_URL')
+    BASKET_SQS_QUEUE_URL = env('BASKET_SQS_QUEUE_URL')
+
+    # Pontoon settings
```
RE: Thunderbird, I've ported that functionality over to specific configuration classes for Thunderbird, rather than a bool flag.

cadecairos commented on donate/settings/thunderbird.py:
```diff
@@ -0,0 +1,6 @@
+
+class ThunderbirdOverrides(object):
+    INSTALLED_APPS = ['donate.thunderbird']
+    TEMPLATES_DIR = [app('thunderbird/templates')]
```
We don't need to include a boolean, since using the Thunderbird configurations implicitly implies that `THUNDERBIRD_INSTANCE` is set to True

cadecairos commented:
remaining todo:

- [ ] Fix default Braintree config
- [ ] Review app webhook env vars in base.py and environment.py
- [ ] Verify review apps are working under this branch

cadecairos commented:
>  Fix default Braintree config

I'm pretty sure what this PR does doesn't change the existing behaviour out of the box. I'd like to handle that in another issue if it's a problem.

Pomax commented on .travis.yml:
```diff
@@ -42,3 +41,4 @@ matrix:
     - nvm use --lts=carbon
     install: "./travis-scripts/npm-install.sh"
     script: npm run test
+
```
😄

Pomax commented on donate/settings/__init__.py:
```diff
@@ -0,0 +1,30 @@
+from .staging import Staging, ThunderbirdStaging
+from .production import Production, ThunderbirdProduction
+
+__all__ = [
```
Django has so much magic in it, I really don't understand why there is no way to just tell it to imexport everything in the dir...

Pomax commented on donate/settings/base.py:
```diff
@@ -0,0 +1,385 @@
+    MEDIA_ROOT = root('media/')
+    MEDIA_URL = '/media/'
+
+    HEROKU_APP_NAME = ''
```
This feels like it wants to be an env var? (with `'''` fallback?)

cadecairos replied:
This is really only ever needed when using the `ReviewApp` configuration, so I can probably drop it from base.py altogether.

cadecairos replied:
yes

Pomax commented on donate/settings/base.py:
```diff
@@ -0,0 +1,385 @@
+    SLACK_WEBHOOK_PONTOON = env('SLACK_WEBHOOK_PONTOON')
+    TRAVIS_LOGS_URL = env('TRAVIS_JOB_WEB_URL', default='')
+
+    @property
```
this is great

Pomax commented on donate/settings/base.py:
```diff
@@ -0,0 +1,385 @@
+        },
+    }
+
+    LANGUAGES = [
```
a nit, but would it be possible to make this a dedicated list in something like `languages.py` and then importing from there?

cadecairos replied:
yeah I'm okay with that.

Pomax commented on donate/settings/development.py:
```diff
@@ -0,0 +1,61 @@
+
+    @classmethod
+    def pre_setup(cls):
+        super(Development, cls).pre_setup()
```
We're in python 3, so the simpler `super().pre_setup()` should work (and I guess the same for any other super(...) calls)

Pomax commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,113 @@
+import environ
+
+CSP_DEFAULT_SRC = ("'self'")
+CSP_SCRIPT_SRC = ("'self'", "'unsafe-inline'", "'unsafe-eval'",
```
Point of concern: do we _need_ `unsafe-inline` and `unsafe-eval`? They're kind of... unsafe

cadecairos replied:
I checked as part of the VA and we determined that we do need at least one of them I'll check which it is.

cadecairos replied:
Based on my tests, we must use unsafe-inline because of Paypal. We can remove unsafe-eval.

Pomax commented on donate/settings/secure.py:
```diff
@@ -0,0 +1,38 @@
+    SECURE_BROWSER_XSS_FILTER = True
+    SECURE_CONTENT_TYPE_NOSNIFF = True
+    SECURE_HSTS_INCLUDE_SUBDOMAINS = True
+    SECURE_HSTS_SECONDS = 60 * 60 * 24 * 31 * 6
```
I think the recommended value is 1 year? (also, fun aside: six months is not 186 days =D)

cadecairos commented on donate/settings/environment.py:
```diff
@@ -0,0 +1,113 @@
+CSP_WORKER_SRC = CSP_DEFAULT_SRC
+
+BRAINTREE_MERCHANT_ACCOUNTS = {
+    'usd': 'usd-ac',
```
I don't even know where this came from. might be what Torchbox had by default

cadecairos commented on donate/settings/braintree.py:
```diff
@@ -0,0 +1,19 @@
+    BRAINTREE_TOKENIZATION_KEY = env('BRAINTREE_TOKENIZATION_KEY')
+    BRAINTREE_MERCHANT_ACCOUNTS = env('BRAINTREE_MERCHANT_ACCOUNTS')
+    BRAINTREE_PLANS = env('BRAINTREE_PLANS')
+    BRAINTREE_MERCHANT_ACCOUNTS_PAYPAL_MICRO = env('BRAINTREE_MERCHANT_ACCOUNTS_PAYPAL_MICRO')
```
`environment.py` defaults this to `{}` so I think so?
